torchtext/csrc/vectors.cpp

```diff
@@ -70,6 +78,213 @@ struct Vectors : torch::CustomClassHolder {
+  std::vector<std::string> tokens;
+  torch::Tensor vectors = torch::zeros({num_lines, vector_dim});
+  std::vector<std::string> dup_tokens;
+  std::unordered_set<std::string> tokens_set;
+  int64_t num_vecs_loaded = 0;
+
+  tokens.reserve(num_lines);
+
+  // concat all loaded tuples
+  for (size_t i = 0; i < tuples.size(); i++) {
+    auto &&subset_tokens = std::move(std::get<0>(tuples[i]));
+    auto &&subset_vectors = std::move(std::get<1>(tuples[i]));
+    auto &&subset_dup_tokens = std::move(std::get<2>(tuples[i]));
+
+    dup_tokens.insert(dup_tokens.end(), subset_dup_tokens.begin(),
+                      subset_dup_tokens.end());
+
+    // process tokens from each tuple
+    for (size_t j = 0; j < subset_tokens.size(); j++) {
+      if (tokens_set.find(subset_tokens[j]) != tokens_set.end()) {
+        dup_tokens.push_back(subset_tokens[j]);
+        continue;
+      }
+
+      tokens_set.insert(subset_tokens[j]);
+      tokens.push_back(subset_tokens[j]);
+      vectors[num_vecs_loaded] = subset_vectors[j];
+      num_vecs_loaded++;
+    }
+  }
+  *out_tuple = std::make_tuple(tokens, vectors.narrow(0, 0, num_vecs_loaded),
+                               dup_tokens);
+}
+
+LoadedVectorsTuple
+_load_token_and_vectors_from_file(const std::string &file_path,
+                                  const int64_t delimiter_ascii = 32,
+                                  const int64_t num_cpus = 10) {
+  std::cout << "Reading file " << file_path << std::endl;
+
+  std::ifstream fin;
+  fin.open(file_path, std::ios::in);
+
+  std::pair<int64_t, int64_t> num_lines_and_vector_dim_pair =
+      _infer_shape(fin, delimiter_ascii);
+  int64_t num_lines = num_lines_and_vector_dim_pair.first;
+  int64_t vector_dim = num_lines_and_vector_dim_pair.second;
+
+  // need chunk size large enough to read entire file
+  int64_t chunk_size = num_lines / num_cpus + 1;
+
+  std::vector<std::future<LoadedVectorsTuple>> futures;
+  std::vector<std::thread> threads;
+  std::vector<LoadedVectorsTuple> tuples;
+
+  // create threads
+  for (int64_t i = 0; i < num_cpus; i++) {
+    std::promise<LoadedVectorsTuple> p;
+    std::future<LoadedVectorsTuple> f = p.get_future();
+    futures.push_back(std::move(f));
+    threads.push_back(
+        std::thread(_load_tokens_from_file_chunk, file_path, i * chunk_size,
+                    std::min(chunk_size, num_lines - (i * chunk_size)),
+                    vector_dim, delimiter_ascii, std::move(p)));
+  }
+
+  // join threads
+  for (int64_t i = 0; i < num_cpus; i++) {
+    threads[i].join();
+  }
+
+  // get all loaded tuples
+  for (int64_t i = 0; i < num_cpus; i++) {
+    tuples.push_back(std::move(futures[i].get()));
+  }
+
+  LoadedVectorsTuple out_tuple;
+  concat_loaded_vectors_tuples(tuples, num_lines, vector_dim, &out_tuple);
+  return out_tuple;
+}
+
+// Registers our custom op with torch.
+TORCH_LIBRARY(torchtext, m) {
+  m.def("_load_token_and_vectors_from_file",
```

Comment:
why do we have underline here?

Reply:
This means it's private, i.e. not supposed to be used directly as a public API.